Firstly, we download necessary libraries and then display the top rows of all tables right away. To remark, we possess six datasets each representing either the USA or Canada and the certain job posting site

In [1]:
import warnings
warnings.filterwarnings('ignore')

In [2]:
import pandas as pd
import numpy as np

In [3]:
canada_glassdoor = pd.read_excel(r"C:\Users\mmatr\OneDrive\Рабочий стол\JobsProject\JobsData\Canada\Glassdoor\Glassdoor_Data analyst_Jobs_Canada.xlsx")
canada_indeed = pd.read_excel(r"C:\Users\mmatr\OneDrive\Рабочий стол\JobsProject\JobsData\Canada\Indeed\Indeed_Job listing_Canada.xlsx")
canada_linkedin = pd.read_excel(r"C:\Users\mmatr\OneDrive\Рабочий стол\JobsProject\JobsData\Canada\LinkedIn\LinkedIn_Data Analyst_jobs_Canada_main.xlsx")
#USA_glassdoor = pd.read_excel(r"C:\Users\mmatr\OneDrive\Рабочий стол\JobsProject\JobsData\USA\GlassDoor\Glassdoor_Data_analyst_Jobs_USA.xlsx")
#USA_indeed = pd.read_excel(r"C:\Users\mmatr\OneDrive\Рабочий стол\JobsProject\JobsData\USA\Indeed\Indeed_USA_Job_listing.xlsx")
#USA_linkedin = pd.read_excel(r"C:\Users\mmatr\OneDrive\Рабочий стол\JobsProject\JobsData\USA\LinkedIn\LinkedIn_Data Analyst_jobs_United_States_main.xlsx")


In [4]:
display(canada_glassdoor.head(2))
display(canada_indeed.head(2))
display(canada_linkedin.head(2))
#display(USA_glassdoor.head(2))
#display(USA_indeed.head(2))
#display(USA_linkedin.head(2))

job_title             company_name  \
0  Data Analyst (New Grad)  THE EMPIRE LIFE INS. CO   
1   Data Reporting Analyst                 PayFacto   

                                         joblink_URL location  \
0  https://www.glassdoor.ca/partner/jobListing.ht...   Canada   
1  https://www.glassdoor.ca/partner/jobListing.ht...  Toronto   

                   salary_range  rating  emp_type  job_level  job_responsib  \
0                           NaN     4.1       NaN        NaN            NaN   
1  $49K - $71K (Glassdoor Est.)     4.1       NaN        NaN            NaN   

   industry  
0       NaN  
1       NaN

job_title company_name  \
0                                Data Analyst     Sun Life   
1  Data Analyst Intern (Technical Operations)    GeoComply   

                                         joblink_URL  \
0  https://ca.indeed.com/rc/clk?jk=aecb38e8a0c8ae...   
1  https://ca.indeed.com/rc/clk?jk=93c24b284d755a...   

                        location  salary_range  rating  emp_type  job_level  \
0  Remote in Toronto, ON M5J 2H7           NaN     3.9       NaN        NaN   
1                    Toronto, ON           NaN     NaN       NaN        NaN   

   job_responsib  industry  
0            NaN       NaN  
1            NaN       NaN

job_title                    company_name  \
0  Data Analyst  LTI - Larsen & Toubro Infotech   
1  Data Analyst                          Altrio   

                                         joblink_URL  \
0  https://ca.linkedin.com/jobs/view/data-analyst...   
1  https://ca.linkedin.com/jobs/view/data-analyst...   

                   location  salary_range  rating   emp_type  \
0  Toronto, Ontario, Canada           NaN     NaN  Full-time   
1  Toronto, Ontario, Canada           NaN     NaN  Full-time   

          job_level           job_responsib  \
0  Mid-Senior level  Information Technology   
1    Not Applicable                 Analyst   

                               industry  
0         IT Services and IT Consulting  
1  Technology, Information and Internet

Now when we ensured that dataframes columns match we can concatenate those into two separate tables based on the country

In [5]:
#usa_jobs = pd.concat([USA_glassdoor, USA_indeed, USA_linkedin])
canada_jobs = pd.concat([canada_glassdoor, canada_indeed, canada_linkedin])

Straightaway we secure our data from being mistakenly repeated and we apply drop_duplicates method subsetting by 'job_title', 'company_name', 'location' columns, so no the same position in the same company from the identical location is displayed, in order to preserve the fairness and quality of future analysis

In [6]:
#usa_jobs = usa_jobs.drop_duplicates(subset = ['job_title', 'company_name'])
canada_jobs = canada_jobs.drop_duplicates(subset = ['job_title', 'company_name', 'location'])

In [7]:
canada_jobs.shape

(1581, 10)

First of all, we will take a look at data type of each column, verify its correctness and glance at overall description:

In [8]:
#display(usa_jobs.dtypes)
canada_jobs.dtypes

job_title         object
company_name      object
joblink_URL       object
location          object
salary_range      object
rating           float64
emp_type          object
job_level         object
job_responsib     object
industry          object
dtype: object

In [9]:
canada_jobs.describe(include='all')

job_title company_name  \
count           1581         1581   
unique           946          899   
top     Data Analyst      TD Bank   
freq             137           22   
mean             NaN          NaN   
std              NaN          NaN   
min              NaN          NaN   
25%              NaN          NaN   
50%              NaN          NaN   
75%              NaN          NaN   
max              NaN          NaN   

                                              joblink_URL     location  \
count                                                1581         1581   
unique                                               1581          303   
top     https://www.glassdoor.ca/partner/jobListing.ht...  Toronto, ON   
freq                                                    1          204   
mean                                                  NaN          NaN   
std                                                   NaN          NaN   
min                                                   NaN          NaN   
25%                                                   NaN          NaN   
50%                                                   NaN          NaN   
75%                                                   NaN          NaN   
max                                                   NaN          NaN   

                       salary_range      rating   emp_type    job_level  \
count                           147  911.000000        441          448   
unique                          136         NaN          5            6   
top     $50K - $60K (Employer Est.)         NaN  Full-time  Entry level   
freq                              4         NaN        357          177   
mean                            NaN    3.718990        NaN          NaN   
std                             NaN    0.422475        NaN          NaN   
min                             NaN    1.300000        NaN          NaN   
25%                             NaN    3.500000        NaN          NaN   
50%                             NaN    3.800000        NaN          NaN   
75%                             NaN    4.000000        NaN          NaN   
max                             NaN    5.000000        NaN          NaN   

                 job_responsib                              industry  
count                      441                                   441  
unique                      24                                    55  
top     Information Technology  Technology, Information and Internet  
freq                       308                                    61  
mean                       NaN                                   NaN  
std                        NaN                                   NaN  
min                        NaN                                   NaN  
25%                        NaN                                   NaN  
50%                        NaN                                   NaN  
75%                        NaN                                   NaN  
max                        NaN                                   NaN

In [10]:
canada_jobs[canada_jobs['company_name'] == 'TD Bank']

job_title company_name  \
2                               Data Analyst Associate      TD Bank   
3                            Data Analyst Intern/Co-op      TD Bank   
64                       Business Analyst Intern/Co-op      TD Bank   
7                               Data Analyst Associate      TD Bank   
84                           Data Analyst Intern/Co-op      TD Bank   
116                          Business Insights Analyst      TD Bank   
117                      Business Analyst Intern/Co-op      TD Bank   
129                                IT Data Analyst III      TD Bank   
176                                  Inventory Analyst      TD Bank   
354                       Sr. Data Analyst (Mainframe)      TD Bank   
468                       Business Systems Analyst III      TD Bank   
510                                Sr. IT Data Analyst      TD Bank   
560  Business Intelligence & Reporting (Undergraduate)      TD Bank   
592                         TDS Operations Analyst III      TD Bank   
606                       Sr Change Management Analyst      TD Bank   
702     Business Intelligence & Reporting Intern/Co-op      TD Bank   
729         Sr. Analyst, Digital Advertising Analytics      TD Bank   
772                       Sr. Business Systems Analyst      TD Bank   
785    Modelling / Forecasting Senior Analyst (Quants)      TD Bank   
815                                      BI Specialist      TD Bank   
877                      Business Info Mgmt Analyst II      TD Bank   
947                   Senior Analyst, Campaign Analyst      TD Bank   

                                           joblink_URL  \
2    https://www.glassdoor.ca/partner/jobListing.ht...   
3    https://www.glassdoor.ca/partner/jobListing.ht...   
64   https://www.glassdoor.ca/partner/jobListing.ht...   
7    https://ca.indeed.com/rc/clk?jk=fd79ddaee38bea...   
84   https://ca.indeed.com/rc/clk?jk=04cd3f49dec8ed...   
116  https://ca.indeed.com/rc/clk?jk=eb018603a6a994...   
117  https://ca.indeed.com/rc/clk?jk=a1b4389650d945...   
129  https://ca.indeed.com/rc/clk?jk=b54a0427de5466...   
176  https://ca.indeed.com/rc/clk?jk=4a39c480a16400...   
354  https://ca.indeed.com/rc/clk?jk=8abefb22dd7ec2...   
468  https://ca.indeed.com/rc/clk?jk=d9b1562dc76ef2...   
510  https://ca.indeed.com/rc/clk?jk=ba8526d1cc32b5...   
560  https://ca.indeed.com/rc/clk?jk=84769819303a76...   
592  https://ca.indeed.com/rc/clk?jk=5b5d9d26f53084...   
606  https://ca.indeed.com/rc/clk?jk=366ea1617021d9...   
702  https://ca.indeed.com/rc/clk?jk=7100af9c967e36...   
729  https://ca.indeed.com/rc/clk?jk=3cd975392e591b...   
772  https://ca.indeed.com/rc/clk?jk=90a46547d1b1c0...   
785  https://ca.indeed.com/rc/clk?jk=c41e6719f3bee7...   
815  https://ca.indeed.com/rc/clk?jk=a4c9a71b430e42...   
877  https://ca.indeed.com/rc/clk?jk=297cef20e8705b...   
947  https://ca.indeed.com/rc/clk?jk=22052827d998b9...   

                         location                  salary_range  rating  \
2                         Toronto  $56K - $95K (Glassdoor Est.)     4.0   
3                         Toronto                           NaN     4.0   
64                        Toronto                           NaN     4.0   
7                     Toronto, ON                           NaN     3.8   
84   Hybrid remote in Toronto, ON                           NaN     3.8   
116                   Toronto, ON                           NaN     3.8   
117         Remote in Toronto, ON                           NaN     3.8   
129                   Toronto, ON                           NaN     3.8   
176                   Toronto, ON                           NaN     3.8   
354                    London, ON                           NaN     3.8   
468                   Toronto, ON                           NaN     3.8   
510      Toronto, ON\n+1 location                           NaN     3.8   
560                   Toronto, ON                           NaN     3.8   
592                   Toro

Splendid, types of all columns matches what we want them to perform. The exception is 'salary_range' field, but the whole section will be devoted for its normalizing.

The 'describe' method also did not bring any ambiguity. The amount of unique titles and company names differs from their total amount, because there are cases when one company offers different jobs. For example TD Bank, the most frequent value, is looking for staff to fill 19 distinct positions.

Let's begin defining the site the job offer was posted, relying on the fields we have in access

In [11]:
#deriving the jobsite, based on joblink_URL column values
#usa_jobs['site'] = usa_jobs['joblink_URL'].apply(lambda x: 'linkedin' if 'linkedin' in str(x).lower() else ('indeed' if 'indeed' in str(x).lower() else 'glassdoor')) 
canada_jobs['site'] = canada_jobs['joblink_URL'].apply(lambda x: 'linkedin' if 'linkedin' in str(x).lower() else ('indeed' if 'indeed' in str(x).lower() else 'glassdoor'))         
        

I suggest to proceed with working out the 'location' field:

In [12]:
#complete 'remote' feature engineering
canada_jobs['remote_locat'] = canada_jobs['location'].apply(lambda x: 'yes' if 'remote' in x.lower() else 'no')
canada_jobs['remote_job_tit'] = canada_jobs['job_title'].apply(lambda x: 'yes' if 'remote' in x.lower() else 'no')
canada_jobs['remote'] = canada_jobs.apply(lambda x: 'yes' if x['remote_job_tit'] == 'yes' else ('yes' if x['remote_locat'] == 'yes' else 'no'), axis=1)
canada_jobs.drop(columns = ['remote_locat', 'remote_job_tit'], axis=1,inplace=True)
canada_jobs[25:70]

job_title  \
25                              Data Analyst - Remote   
26                        Analyst, Data and Reporting   
27                                       Data Analyst   
28                              Data Analyst - Remote   
29                                       Data Analyst   
30                                       Data Analyst   
31                          Analyst, Data & Analytics   
32  System Planner Data Analyst Population Health,...   
33                     Data Analyst - Crime Analytics   
34                            Business Data Analyst I   
35                                       Data Analyst   
36           Data Analyst Co-op Student (Spring Term)   
37                                       Data Analyst   
38                                       Data Analyst   
39                                       Data Analyst   
40  Data & Analytics - Associate Data Analyst (May...   
41                                       Data Analyst   
42         Information Analyst - Data Quality (09SID)   
43                                       Data Analyst   
44                           Systems and Data Analyst   
45           IP Procurement Data Analyst co-op/intern   
46                                Senior Data Analyst   
47                       Senior Data Analyst, Product   
48                                       Data Analyst   
49                                       Data Analyst   
50                   Data Analyst (Power BI) - REMOTE   
51                                       Data Analyst   
52                                       Data Analyst   
53                                       Data Analyst   
54               Business Analyst (12 month contract)   
55    Data Analyst & Business Intelligence Specialist   
56                                       Data Analyst   
57                 Business Analyst - BI Data Analyst   
58                                       Data Analyst   
59  Product Analyst, Summer 2023 (Co-op/Internship...   
60                               Data Quality Analyst   
61                                       Data Analyst   
62                                       Data Analyst   
63                         Business Analyst (Payment)   
64                      Business Analyst Intern/Co-op   
65                                       Data Analyst   
66                                       Data Analyst   
67                         Sr Data Governance Analyst   
68                                       Data Analyst   
69             Data and Business Intelligence Analyst   

                            company_name  \
25                                  KOHO   
26       Dignity Health Global Education   
27        Synergee Fitness Worldwide Inc   
28                                  Humi   
29               Carrier Sekani Family S   
30             Strathcona Resources Ltd.   
31                            Air Canada   
32                     Oak Valley Health   
33                      Canadian Pacific   
34  Trillium Supply Chain | Supply Chain   
35                              Sun Life   
36               Ridley College (Canada)   
37                    Annex IT Solutions   
38                         Winobell Inc.   
39                  Environics Analytics   
40                               Aritzia   
41    Techtronic Industries Canada, Inc.   
42                  eHealth Saskatchewan   
43                               Newrest   
44                     Strathcona County   
45                                 NOKIA   
46            Government of Saskatchewan   
47                                Docker   
48                               Infosys   
49                            Indellient   
50                Software International   
51         IPNET INSTITUTE OF TECHNOLOGY   
52                    Future Electronics   
53  Workplace Safety and Insurance Board   
54                                Nestle   
55                        Branksome Hall   
56           

In [13]:
#pull out the city and state
canada_jobs['city'] = canada_jobs['location'].apply(lambda x:x.split('\n')[0].split(',')[0])
canada_jobs['state'] = canada_jobs['location'].apply(lambda x: x.split('\n')[0].split(',')[1][:3] if ',' in x else 'not indicated')

In [14]:
canada_jobs[25:70]

job_title  \
25                              Data Analyst - Remote   
26                        Analyst, Data and Reporting   
27                                       Data Analyst   
28                              Data Analyst - Remote   
29                                       Data Analyst   
30                                       Data Analyst   
31                          Analyst, Data & Analytics   
32  System Planner Data Analyst Population Health,...   
33                     Data Analyst - Crime Analytics   
34                            Business Data Analyst I   
35                                       Data Analyst   
36           Data Analyst Co-op Student (Spring Term)   
37                                       Data Analyst   
38                                       Data Analyst   
39                                       Data Analyst   
40  Data & Analytics - Associate Data Analyst (May...   
41                                       Data Analyst   
42         Information Analyst - Data Quality (09SID)   
43                                       Data Analyst   
44                           Systems and Data Analyst   
45           IP Procurement Data Analyst co-op/intern   
46                                Senior Data Analyst   
47                       Senior Data Analyst, Product   
48                                       Data Analyst   
49                                       Data Analyst   
50                   Data Analyst (Power BI) - REMOTE   
51                                       Data Analyst   
52                                       Data Analyst   
53                                       Data Analyst   
54               Business Analyst (12 month contract)   
55    Data Analyst & Business Intelligence Specialist   
56                                       Data Analyst   
57                 Business Analyst - BI Data Analyst   
58                                       Data Analyst   
59  Product Analyst, Summer 2023 (Co-op/Internship...   
60                               Data Quality Analyst   
61                                       Data Analyst   
62                                       Data Analyst   
63                         Business Analyst (Payment)   
64                      Business Analyst Intern/Co-op   
65                                       Data Analyst   
66                                       Data Analyst   
67                         Sr Data Governance Analyst   
68                                       Data Analyst   
69             Data and Business Intelligence Analyst   

                            company_name  \
25                                  KOHO   
26       Dignity Health Global Education   
27        Synergee Fitness Worldwide Inc   
28                                  Humi   
29               Carrier Sekani Family S   
30             Strathcona Resources Ltd.   
31                            Air Canada   
32                     Oak Valley Health   
33                      Canadian Pacific   
34  Trillium Supply Chain | Supply Chain   
35                              Sun Life   
36               Ridley College (Canada)   
37                    Annex IT Solutions   
38                         Winobell Inc.   
39                  Environics Analytics   
40                               Aritzia   
41    Techtronic Industries Canada, Inc.   
42                  eHealth Saskatchewan   
43                               Newrest   
44                     Strathcona County   
45                                 NOKIA   
46            Government of Saskatchewan   
47                                Docker   
48                               Infosys   
49                            Indellient   
50                Software International   
51         IPNET INSTITUTE OF TECHNOLOGY   
52                    Future Electronics   
53  Workplace Safety and Insurance Board   
54                                Nestle   
55                        Branksome Hall   
56           

Still there are some misperfections in our own derived fields, such as 'remote' and 'Canada' in 'city' column, however, these values are quite informative and can be handled easily during the exploratory analysis and visualization phases. Therefore we are done we these attributes

In [15]:
canada_jobs.head()

job_title             company_name  \
0    Data Analyst (New Grad)  THE EMPIRE LIFE INS. CO   
1     Data Reporting Analyst                 PayFacto   
2     Data Analyst Associate                  TD Bank   
3  Data Analyst Intern/Co-op                  TD Bank   
4     Data Analyst Part time        TGT Solutions Inc   

                                         joblink_URL location  \
0  https://www.glassdoor.ca/partner/jobListing.ht...   Canada   
1  https://www.glassdoor.ca/partner/jobListing.ht...  Toronto   
2  https://www.glassdoor.ca/partner/jobListing.ht...  Toronto   
3  https://www.glassdoor.ca/partner/jobListing.ht...  Toronto   
4  https://www.glassdoor.ca/partner/jobListing.ht...   Remote   

                              salary_range  rating emp_type job_level  \
0                                      NaN     4.1      NaN       NaN   
1             $49K - $71K (Glassdoor Est.)     4.1      NaN       NaN   
2             $56K - $95K (Glassdoor Est.)     4.0      NaN       NaN   
3                                      NaN     4.0      NaN       NaN   
4  $75.00 - $85.00 Per Hour(Employer Est.)     4.0      NaN       NaN   

  job_responsib industry       site remote     city          state  
0           NaN      NaN  glassdoor     no   Canada  not indicated  
1           NaN      NaN  glassdoor     no  Toronto  not indicated  
2           NaN      NaN  glassdoor     no  Toronto  not indicated  
3           NaN      NaN  glassdoor     no  Toronto  not indicated  
4           NaN      NaN  glassdoor    yes   Remote  not indicated

Next stage is to get familiar with salary_range values:

In [16]:
canada_jobs['salary_range'].unique()

array([nan, '$49K - $71K (Glassdoor Est.)',
       '$56K - $95K (Glassdoor Est.)',
       '$75.00 - $85.00 Per Hour(Employer Est.)',
       '$44K - $80K (Glassdoor Est.)', '$50K - $55K (Employer Est.)',
       '$77K - $100K (Glassdoor Est.)', '$68K - $87K (Glassdoor Est.)',
       '$24.04 - $26.44 Per Hour(Employer Est.)',
       '$69K - $81K (Glassdoor Est.)',
       '$38.99 - $47.46 Per Hour(Employer Est.)',
       '$52K - $70K (Glassdoor Est.)', '$60K - $81K (Glassdoor Est.)',
       '$81K - $85K (Glassdoor Est.)', '$58K - $75K (Glassdoor Est.)',
       '$78K - $85K (Glassdoor Est.)', '$85K - $90K (Employer Est.)',
       '$64K - $68K (Glassdoor Est.)',
       '$41.17 - $51.60 Per Hour(Employer Est.)',
       '$60K - $70K (Glassdoor Est.)', '$50K - $60K (Employer Est.)',
       '$35.86 - $44.12 Per Hour(Employer Est.)',
       '$43K - $80K (Glassdoor Est.)', '$73K - $77K (Glassdoor Est.)',
       '$43.27 - $53.14 Per Hour(Employer Est.)',
       '$89K - $111K (Glassdoor Est.)', '$57

It is necessary to mark certain rows with no salary provided, so we are able to slice the data, based on this predicate

In [17]:
canada_jobs['salary_range'] = canada_jobs['salary_range'].fillna('-1')

As we can see there are two noticeable patterns the US table and one in Canada's. Thus, I suggest to retrieve the meaningful values from the jobs in the country, whose symbol is clover, at first.

At first, we filter out the rows, where salary is not provided, then replace all capital letters "K" after the number with three zeroes to indicate thousands, remove dollar sign before actual numbers and truncate the text part at all

In [18]:
clean_salary_canada = canada_jobs[canada_jobs['salary_range'] != '-1']

In [19]:
#Ensure there is no missing values
clean_salary_canada['salary_range'].isna().sum()

0

In [20]:
clean_salary_canada['salary_range'] = clean_salary_canada['salary_range'].apply(lambda x:x.replace('K', '000').replace('$', '').split('(')[0])

Now we get a dummy variable to indicate the offerings, where salaries are shown per hour . It will help us in future calculations

In [21]:
clean_salary_canada['per_hour_variable'] = clean_salary_canada['salary_range'].apply(lambda x: 1 if 'per hour' in str(x).lower() else 0)

Now we can get rid of that string as well:

In [22]:
clean_salary_canada['salary_range'] = clean_salary_canada['salary_range'].apply(lambda x:x.split(' P')[0])

Finally, we are able to split the range into minimal and maximum wage per year(we will convert hour rates later):

In [23]:
#clean_salary_canada['salary_range'] = clean_salary_canada['salary_range'].replace('nan', 'none - none')
#canada_jobs['salary_range'] = canada_jobs['salary_range'].replace(-1, 'none - none')


In [24]:
clean_salary_canada['min_salary'] = clean_salary_canada['salary_range'].apply(lambda x:x.split('-')[0])

In [25]:
#Looking at unique values of salary range I noticed some employers provided only one certain wage
#So in the process of indentifying maximal salary the code was wrapped in if-condition
for index, row in clean_salary_canada.iterrows(): 
    if '-' in row['salary_range']:
        clean_salary_canada.at[index, 'max_salary'] = row['salary_range'].split('-')[1]
    else:
        clean_salary_canada.at[index, 'max_salary'] = row['min_salary']

In [26]:
clean_salary_canada.head(50)

job_title  \
1                              Data Reporting Analyst   
2                              Data Analyst Associate   
4                              Data Analyst Part time   
7                                        Data Analyst   
8                                 Junior Data Analyst   
9                      IT Business Data Analyst - SQL   
10                                       Data Analyst   
11                Data Analyst Intern (Risk Research)   
12                                 Data Analyst Co-Op   
13                            Data and Report Analyst   
14  Binance Accelerator Programme - Junior Data An...   
15                                       Data Analyst   
16                                       Data Analyst   
18  Intern Analyst, Data & Technology (Summer 2023...   
19         Data Analyst Intern (Technical Operations)   
20                     Big Data Analyst - PL SQL, SQL   
21                                       Data Analyst   
23                                       Data Analyst   
24                        Data and Evaluation Analyst   
26                        Analyst, Data and Reporting   
27                                       Data Analyst   
29                                       Data Analyst   
30                                       Data Analyst   
31                          Analyst, Data & Analytics   
32  System Planner Data Analyst Population Health,...   
33                     Data Analyst - Crime Analytics   
34                            Business Data Analyst I   
35                                       Data Analyst   
36           Data Analyst Co-op Student (Spring Term)   
37                                       Data Analyst   
38                                       Data Analyst   
39                                       Data Analyst   
40  Data & Analytics - Associate Data Analyst (May...   
41                                       Data Analyst   
42         Information Analyst - Data Quality (09SID)   
43                                       Data Analyst   
44                           Systems and Data Analyst   
45           IP Procurement Data Analyst co-op/intern   
46                                Senior Data Analyst   
48                                       Data Analyst   
50                   Data Analyst (Power BI) - REMOTE   
51                                       Data Analyst   
52                                       Data Analyst   
53                                       Data Analyst   
54               Business Analyst (12 month contract)   
55    Data Analyst & Business Intelligence Specialist   
56                                       Data Analyst   
57                 Business Analyst - BI Data Analyst   
59  Product Analyst, Summer 2023 (Co-op/Internship...   
60                               Data Quality Analyst   

                                  company_name  \
1                                     PayFacto   
2                                      TD Bank   
4                            TGT Solutions Inc   
7   One Care Home & Community Support Services   
8                              Earthscape Play   
9                                         Citi   
10                           CAAT Pension Plan   
11                                   GeoComply   
12               Canada Life Assurance Company   
13                  Government of Saskatchewan   
14                                    Humanity   
15                       Rubicon Organics Inc.   
16                                    Manulife   
18                     University Pension Plan   
19                                   GeoComply   
20                                        Citi   
21                Genisis technology solutions   
23                                   Atco Ltd.   
24                  Government of Saskatchewan   
26             Dignity Health Global Education   
27              Synergee Fitness Worldwide Inc   
29                     Carrier Sekan

At this point we shall declare newly created columns as of float data type:

In [27]:
clean_salary_canada['min_salary'] = clean_salary_canada['min_salary'].astype('float')
clean_salary_canada['max_salary'] = clean_salary_canada['max_salary'].astype('float')

Now we are going to turn all values into salary per year. That is where 'per hour' variable will come in handy

In [28]:
#According to the source(https://snappyrates.ca/working-days-in-canada/), 
#there are 173 working hours per month on average, consequently 2080 hours
#this number will be taken in the computation
for index, row in clean_salary_canada.iterrows(): 
    if row['per_hour_variable'] == 1:
        clean_salary_canada.at[index, 'min_salary'] = row['min_salary']*2080
        clean_salary_canada.at[index, 'max_salary'] = row['max_salary']*2080

Additionally, while parsing the unique values of saly range field, I saw  that several records contained estimated salary per month next couple of lines of code are going to fix the issue


In [29]:
#We just multiply the salary on the amount of months in a year
for index, row in clean_salary_canada.iterrows(): 
    if (row['min_salary'] >= 1000) and (row['min_salary'] <= 15000):
        clean_salary_canada.at[index, 'min_salary'] = row['min_salary']*12
        clean_salary_canada.at[index, 'max_salary'] = row['max_salary']*12



Let's check the result of the manipulation:

In [30]:
display(clean_salary_canada['max_salary'].describe())
clean_salary_canada['min_salary'].describe()

count       147.000000
mean      83516.429932
std       23424.174060
min       34008.000000
25%       66000.000000
50%       81000.000000
75%       96000.000000
max      176800.000000
Name: max_salary, dtype: float64

count       147.000000
mean      63386.231293
std       16981.766860
min       32240.000000
25%       50001.600000
50%       61000.000000
75%       73118.400000
max      156000.000000
Name: min_salary, dtype: float64

Everything looks just fine as it is supposed to do, except maximum value it seems strange. We should examine the record with that salary and apply for that job immediately:

In [31]:
#Just in case, we drop duplicates one more time to prevent any redundancy
clean_salary_canada = clean_salary_canada.drop_duplicates(subset = ['job_title', 'company_name'])

In [32]:
#clean_salary_canada.loc[clean_salary_canada['max_salary'] == np.max(clean_salary_canada['max_salary'])]
clean_salary_canada.sort_values('max_salary', ascending = False).head(10)

job_title  \
4                               Data Analyst Part time   
51                                        Data Analyst   
77                              Data Engineer/ Analyst   
129                                       Data Analyst   
37                                        Data Analyst   
45            IP Procurement Data Analyst co-op/intern   
72                                        Data Analyst   
63                          Business Analyst (Payment)   
118               Business Analyst (wealth management)   
171  Analyste des rapports de données - Data Report...   

                      company_name  \
4                TGT Solutions Inc   
51   IPNET INSTITUTE OF TECHNOLOGY   
77      Stuvi Consultancy Services   
129                        Mphasis   
37              Annex IT Solutions   
45                           NOKIA   
72                   Brik Partners   
63                     Snappyhires   
118                    Snappyhires   
171                       PayFacto   

                                           joblink_URL  location  \
4    https://www.glassdoor.ca/partner/jobListing.ht...    Remote   
51   https://www.glassdoor.ca/partner/jobListing.ht...  Montreal   
77   https://www.glassdoor.ca/partner/jobListing.ht...   Toronto   
129  https://www.glassdoor.ca/partner/jobListing.ht...  Edmonton   
37   https://www.glassdoor.ca/partner/jobListing.ht...    Remote   
45   https://www.glassdoor.ca/partner/jobListing.ht...    Ottawa   
72   https://www.glassdoor.ca/partner/jobListing.ht...   Toronto   
63   https://www.glassdoor.ca/partner/jobListing.ht...   Toronto   
118  https://www.glassdoor.ca/partner/jobListing.ht...   Toronto   
171  https://www.glassdoor.ca/partner/jobListing.ht...  Montreal   

         salary_range  rating emp_type job_level job_responsib industry  \
4       75.00 - 85.00     4.0      NaN       NaN           NaN      NaN   
51      20.21 - 82.28     NaN      NaN       NaN           NaN      NaN   
77      50.00 - 70.00     NaN      NaN       NaN           NaN      NaN   
129   81000 - 131000      3.7      NaN       NaN           NaN      NaN   
37      19.21 - 60.71     NaN      NaN       NaN           NaN      NaN   
45   108000 - 125000      4.2      NaN       NaN           NaN      NaN   
72      30.00 - 60.00     NaN      NaN       NaN           NaN      NaN   
63    68000 - 122000      4.9      NaN       NaN           NaN      NaN   
118   68000 - 122000      4.9      NaN       NaN           NaN      NaN   
171   60000 - 121000      4.1      NaN       NaN           NaN      NaN   

          site remote      city          state  per_hour_variable  min_salary  \
4    glassdoor    yes    Remote  not indicated                  1    156000.0   
51   glassdoor     no  Montreal  not indicated                  1     42036.8   
77   glassdoor     no   Toronto  not indicated                  1    104000.0   
129  glassdoor     no  Edmonton  not indicated                  0     81000.0   
37   glassdoor    yes    Remote  not indicated                  1     39956.8   
45   glassdoor     no    Ottawa  not indicated                  0    108000.0   
72   glassdoor     no   Toronto  not indicated                  1     62400.0   
63   glassdoor     no   Toronto  not indicated                  0     68000.0   
118  glassdoor     no   Toronto  not indicated                  0     68000.0   
171  glassdoor     no  Montreal  not indicated                  0     60000.0   

     max_salary  
4      176800.0  
51     171142.4  
77     145600.0  
129    131000.0  
37     126276.8  
45     125000.0  
72     124800.0  
63     122000.0  
118    122000.0  
171    121000.0

It still looks weird, but we see there is a phrase 'part-time' in the title of the job. Thus, supposedly the calculated maximum and minimal wage are hard to reach, but we chose the amount of working hours to be met by full-time positions(2080 hours per year). For the part-time roles this number is equivocal. The authorized source Indeed(https://ca.indeed.com/career-advice/finding-a-job/how-many-hours-is-part-time) states that in Canada any amount of hours worked less than 30 is considered as part-time. Glassdoor(https://www.glassdoor.com/blog/guide/how-many-hours-is-part-time/), however, defines a quite wide but more certain range - 20 to 29 hours. We are going to choose the middle number, which, in this occassion, is 25 hours. Some of not hard maths and we get 1300 hours per year. This variable will be used in part-time salary calculations.

In [33]:
for index, row in clean_salary_canada.iterrows():
    if  ('part time' in row['job_title'].lower()) or ('part-time' in row['job_title'].lower()):
        clean_salary_canada.at[index, 'min_salary'] = (row['min_salary']/2080)*1300 #we divide by 2080 to get back to hour rate
        clean_salary_canada.at[index, 'max_salary'] = (row['max_salary']/2080)*1300

And to make salary attribute comparable with the USA jobs data, we convert the salaries into US dollars:

In [34]:
#currency rate on 01.02.2023: 1 CAD = 0,74356 USD
clean_salary_canada['min_salary'] = clean_salary_canada['min_salary']*0.74
clean_salary_canada['max_salary'] = clean_salary_canada['max_salary']*0.74

In [35]:
clean_salary_canada.sort_values('max_salary', ascending=False).head()

job_title                   company_name  \
51                               Data Analyst  IPNET INSTITUTE OF TECHNOLOGY   
77                     Data Engineer/ Analyst     Stuvi Consultancy Services   
129                              Data Analyst                        Mphasis   
37                               Data Analyst             Annex IT Solutions   
45   IP Procurement Data Analyst co-op/intern                          NOKIA   

                                           joblink_URL  location  \
51   https://www.glassdoor.ca/partner/jobListing.ht...  Montreal   
77   https://www.glassdoor.ca/partner/jobListing.ht...   Toronto   
129  https://www.glassdoor.ca/partner/jobListing.ht...  Edmonton   
37   https://www.glassdoor.ca/partner/jobListing.ht...    Remote   
45   https://www.glassdoor.ca/partner/jobListing.ht...    Ottawa   

         salary_range  rating emp_type job_level job_responsib industry  \
51      20.21 - 82.28     NaN      NaN       NaN           NaN      NaN   
77      50.00 - 70.00     NaN      NaN       NaN           NaN      NaN   
129   81000 - 131000      3.7      NaN       NaN           NaN      NaN   
37      19.21 - 60.71     NaN      NaN       NaN           NaN      NaN   
45   108000 - 125000      4.2      NaN       NaN           NaN      NaN   

          site remote      city          state  per_hour_variable  min_salary  \
51   glassdoor     no  Montreal  not indicated                  1   31107.232   
77   glassdoor     no   Toronto  not indicated                  1   76960.000   
129  glassdoor     no  Edmonton  not indicated                  0   59940.000   
37   glassdoor    yes    Remote  not indicated                  1   29568.032   
45   glassdoor     no    Ottawa  not indicated                  0   79920.000   

     max_salary  
51   126645.376  
77   107744.000  
129   96940.000  
37    93444.832  
45    92500.000

In [36]:
cleaned_canada_df = canada_jobs.merge(clean_salary_canada, on=['job_title','company_name'], how='outer', suffixes=['', '_clean'])

In [37]:
columns_to_drop = ['joblink_URL_clean', 'location_clean',
       'salary_range_clean', 'rating_clean', 'emp_type_clean',
       'job_level_clean', 'job_responsib_clean', 'industry_clean',
       'site_clean', 'remote_clean', 'city_clean', 'state_clean', 'per_hour_variable']
cleaned_canada_df = cleaned_canada_df.drop(columns_to_drop, axis=1)

In [38]:
cleaned_canada_df.columns

Index(['job_title', 'company_name', 'joblink_URL', 'location', 'salary_range',
       'rating', 'emp_type', 'job_level', 'job_responsib', 'industry', 'site',
       'remote', 'city', 'state', 'min_salary', 'max_salary'],
      dtype='object')

In [39]:
#Added column which indicates what country is the job from for the convenience's sake
#during future union with the USA's jobs dataset
cleaned_canada_df['Country'] = 'Canada'

With intent to clean the data from copies across the postings from different websites, we drop duplicates through 'job_title', 'company_name', 'city' columns.

In [40]:
cleaned_canada_df = cleaned_canada_df.drop_duplicates(subset= ['job_title', 'company_name', 'city'])
cleaned_canada_df.shape

(1476, 17)

We have a dataset consisted of 1476 cleaned and verified records about jobs from all over Canada ready to be exported.

In [43]:
cleaned_canada_df.to_csv('C:/Users/mmatr/OneDrive/Рабочий стол/JobsProject/Cleaned_DFs/ready_canadaJobs_df')